singularity-eos/eos/eos_eospac.hpp

```diff
@@ -72,103 +72,1156 @@ class EOSPAC : public EosBase<EOSPAC> {
-  // inline
-  // void BulkModulusFromDensityInternalEnergy(ConstRealIndexer &&rhos,
-  //                                           ConstRealIndexer &&sies,
-  //                                           RealIndexer &&bmods,
-  //                                           const int num,
-  //                                           LambdaIndexer &&lambdas) const;
-
-  // template<typename RealIndexer, typename ConstRealIndexer, typename LambdaIndexer>
-  // inline
-  // void GruneisenParamFromDensityTemperature(ConstRealIndexer &&rhos,
-  //                                           ConstRealIndexer &&temperatures,
-  //                                           RealIndexer &&gm1s,
-  //                                           const int num,
-  //                                           LambdaIndexer &&lambdas) const;
-
-  // template<typename RealIndexer, typename ConstRealIndexer, typename LambdaIndexer>
-  // inline
-  // void GruneisenParamFromDensityInternalEnergy(ConstRealIndexer &&rhos,
-  //                                              ConstRealIndexer &&sies,
-  //                                              RealIndexer &&gm1s,
-  //                                              const int num,
-  //                                              LambdaIndexer &&lambdas) const;
-
-  // template<typename RealIndexer, typename LambdaIndexer>
-  // inline
-  // void FillEos(RealIndexer &&rhos, RealIndexer &&temps, RealIndexer &&energies,
-  //              RealIndexer &&presses, RealIndexer &&cvs, RealIndexer &&bmods,
-  //              const int num, const unsigned long output,
-  //              LambdaIndexer &&lambdas) const;
+  // EOSPAC vector implementations
+  template <int CHUNK_SIZE = 64, typename RealIndexer, typename ConstRealIndexer,
+            typename LambdaIndexer>
+  inline void
+  TemperatureFromDensityInternalEnergy(ConstRealIndexer &&rhos, ConstRealIndexer &&sies,
+                                       RealIndexer &&temperatures, const int num,
+                                       LambdaIndexer &&lambdas) const {
+    using namespace EospacWrapper;
+    int nchunks = (num / CHUNK_SIZE) + (num % CHUNK_SIZE ? 1 : 0);
+    std::array<EOS_REAL, CHUNK_SIZE> R, E, T, dTdr, dTde;
+
+    for (int ichunk = 0; ichunk < nchunks; ++ichunk) {
+      int istart = ichunk * CHUNK_SIZE;
+      int iend = std::min(num, istart + CHUNK_SIZE);
+      int nelements = iend - istart;
+
+      for (int i = 0; i < nelements; ++i) {
+        int j = istart + i;
+        R[i] = rhos[j];
+        E[i] = sieToSesame(sies[j]);
+      }
+
+      EOS_INTEGER table = TofRE_table_;
+      eosSafeInterpolate(&table, nelements, R.data(), E.data(), T.data(), dTdr.data(),
+                         dTde.data(), "TofRE", Verbosity::Quiet);
+
+      for (int i = 0; i < nelements; ++i) {
+        int j = istart + i;
+        temperatures[j] = temperatureFromSesame(T[i]);
+      }
+    }
+  }
```

Comment:
Have you tested the chunk size version of the code on GPUs? I'm pretty sure as written, this will not work, as the chunk are filled on CPU, whereas the data may live on GPU. This might be fixable with synchronized memory or with explicit deep copies, but it's something to be aware/concerned about.